server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/git/rocksdb/EncryptionGitStorage.java

```diff
@@ -94,18 +97,27 @@ ObjectId insertObject(ObjectId objectId, int type, byte[] data, int off, int len
         }
 
         final byte[] nonce = AesGcmSivCipher.generateNonce();
-        final byte[] nonceAndType = new byte[NONCE_SIZE_BYTES + 4];
+        // Generate a new DEK for the data so that we don't decrypt and encrypt the data again when the
+        // repository key is rotated.
```

Comment:
Implementation looks fine. Just a couple of questions:
- Does a repository's DEK remain unchanged after it is created?
- If we use DEK per object, do you think rewrapping achieves the same goal as key rotate?

Reply:
> Does a repository's DEK remain unchanged after it is created?
If we use DEK per object, do you think rewrapping achieves the same goal as key rotate?

The repository's DEK is rotated. When the repository's DEK is rotated, the old repository's DEK decrypts this object DEK and the new repository's DEK re-encrypts this object DEK. The object DEK is never rotated and it's only used once for the corresponding object. This is used in envelope-encryption:
https://jayendrapatil.com/tag/envelope-encryption/

Reply:
@ikhoon Would you mind reviewing my answer?

Reply:
> When the repository's DEK is rotated, the old repository's DEK decrypts this object DEK and the new repository's DEK re-encrypts this object DEK.

Understood. 

Thanks for the link. I didn't know that it is a common practice to generate DEK for each data item.